lib/surface/compiler/eex_engine.ex

```diff
@@ -23,16 +23,78 @@ defmodule Surface.Compiler.EExEngine do
       context: []
     }
 
-    nodes
-    |> to_token_sequence()
-    |> generate_buffer(state.engine.init(opts), state)
+    required_modules =
+      nodes
+      |> collect_modules_to_require()
```

Comment:
My original intention was runtime performance and compile time completeness/correctness. LiveView treats `requires` as taints that disable change tracking (at least that's my understanding from trolling through their codebase - https://github.com/phoenixframework/phoenix_live_view/blob/master/lib/phoenix_live_view/engine.ex#L879-L884). That being said, I see your point about the cost of traversing the tree multiple times.

I'll make that change as it matches with current behavior, but I think it might be worth investigating how live view treats `require`s and potentially re-evaluating in the future

Reply:
> LiveView treats `requires` as taints that disable change tracking

Oh, damn it. Thanks for the info. In this case we'll have to re-evaluate this very soon. We need to keep change tracking working as expected. If we can't find a way to solve this on our side, then we'll need to talk to the Phoenix core team to find an alternative.

Reply:
I threw together a small app to test out the recompilation with macros (https://github.com/lnr0626/macro-test), and I'm seeing some odd behavior:

After compiling, it appears to have the expected behavior when the file changes:
```
$ touch lib/macro_test_web/live/components/macro.ex && mix compile --verbose
Compiling 2 files (.ex)
Compiled lib/macro_test_web/live/components/macro.ex
Compiled lib/macro_test_web/live/page_live.ex
```

However, when I modify either `page_live.ex` or `macro.ex` in vscode, save it through vs code, and run `mix compile --verbose`, mix doesn't recompile anything, this behavior is the same when using the external co-located template (`page_live.sface`), and when I inline that template into a `render/1` function. When I save `page_live.sface`, it does recompile `page_live.ex`.

Any thoughts on what might be causing that?

Reply:
Are you using ElixirLS? Maybe it's compiling the file already after saving. However, this is weird since, as far as I know, it should recompile it in a separate folder. You can try to take a look at the `Output` panel:

![image](https://user-images.githubusercontent.com/457237/89715667-b213c180-d97d-11ea-9009-2c7670c28254.png)

Reply:
If this is the expected behaviour when using `ElixirLS`, then we can move on :)

Reply:
I am using `ElixirLS`, however it happens even when i disable/remove the extension, so I'm not sure it's related to the extension. I tried editing the file in vim, and it does work as expected. It's definitely something related to vscode specifically though - vscode seems to be touching some files under _build which i assume are related to mix's change detection

Reply:
I tried it on another computer with a fresh install of vscode and it's working as expected (i.e. save a file in vscode, run `mix compile --verbose`, see the expected files are recompiled) - as long as you aren't also seeing that behavior I think we can probably assume it's something specific to that computer/code install, and not specific to this PR

Reply:
@lnr0626 I'm using VS Code with ElixirLS. I've cloned the project and it works as expected so I don't think it's related to this PR at all.

Reply:
@lnr0626 I've just confirmed that adding `<% require Mod %>` does not disable change tracking. LiveView treats `require` as taints only if passed as arguments. 

For instance, the change tracking works perfectly for the following code:

```elixir
def render(assigns) do
  ~L"""
  <% require MyComponent %>
  <%= live_component @socket, MyComponent, id: 2 do %>
    <% require Inner %>
    <%= live_component @socket, Inner, [] do %>
      My Inner
    <% end %>
  <% end %>
  """
end
```
However, if you try this:

```elixir
def render(assigns) do
  ~L"""
  <%= live_component @socket, MyComponent, id: 2, some_prop: require(MyComponent) do %>
    <!--                                                     ^^^^^^^^^^^^^^^^^^^^  -->
    <%= live_component @socket, Inner, [] do %>
      My Inner
    <% end %>
  <% end %>
  """
end
```

Then LV cannot track changes anymore and every single change in the state will trigger a new update for all components.

This is great news since we can stick with the simplest approach and have both, performance and change tracking ;)

Reply:
BTW, regarding change tracking, we still have an issue to solve. The code we generate already disable it. For instance:

```elixir
live_component(
  @socket,
  MyComponent,
  Surface.build_assigns(var!(assigns), ...)
  #                          ^^^^^^^
) do
    ...
end
```

We should never pass `assigns` directly. If we do, LV cannot track changes anymore since it doesn't know which assigns should be tracked for the component. The solution is to pass all context assigns explicitly in the call. 

If you feel this issue may require a lot of work, we can address it later in a separate PR after merging this one.

Reply:
That's awesome news!

I found couple issues with the requires that I'm going to track down first, then I'll look into this. That method only needs the value of `@__surface__[:context]`, however the `__surface__` assign isn't guaranteed to be there - otherwise we could just do `@__surface__[:context]`.

Reply:
Let me know what you think of https://github.com/msaraiva/surface/pull/98/commits/95b9005e481f6137ca5906d9bb77f11dca46a0ac

Reply:
hm, that doesn't quite do it. I'm going to spend a bit more time with it later today

Reply:
The issue with that approach is if you `use Phoenix.LiveView` instead of `Surface.LiveView`, it won't setup that default assign value, which means that just importing `Surface` and using `~H` doesn't just work

Reply:
I guess, for now, we could go with this solution and in case someone needs use a Surface component inside a LV using `use Phoenix.LiveView`, we can say that adding `assign_new(socket, :__surface__, fn -> %{} end)` in `mount/3` is a requirement. We can even wrap the code so the user can do something like `Surface.init(socket)`.

I believe integration with regular LV applications can be addressed separately since it has a broader scope and probably involves a few other issues. WDTY?

Reply:
I think that makes sense. I'll add some info into the readme and expose a `Surface.init(...)` so it's a little more user friendly